File: solutions/cpp/Problem2.cpp
#pragma once
/*
2. Add Two Numbers

You are given two non - empty linked lists representing two non - negative integers.
The digits are stored in reverse order and each of their nodes contain a single digit.
Add the two numbers and return it as a linked list.

You may assume the two numbers do not contain any leading zero, except the number 0 itself.
*/

#include <iostream>
#include "Source.h"


class Solution {
public:
	ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
		ListNode* result = new ListNode(-1);
		ListNode* cur = result;
		int nUp = 0;
		while (l1 != NULL && l2 != NULL)
		{
			int tmp = nUp + l1->val + l2->val;
			nUp = 0;
			cur->next = new ListNode(tmp % 10);
			while (tmp >= 10)
			{
				tmp /= 10;
				nUp += 1;
			}
			cur = cur->next;
			l1 = l1->next;
			l2 = l2->next;
		}
		ListNode* l;
		if (l1 == NULL)
		{
			l = l2;
		}
		else
		{
			l = l1;
		}

		while (l != NULL)
		{
			int tmp = nUp + l->val;
			nUp = 0;
			cur->next = new ListNode(tmp % 10);
			while (tmp >= 10)
			{
				tmp /= 10;
				nUp += 1;
			}
			cur = cur->next;
			l = l->next;
		}

		if (nUp)
		{
			cur->next = new ListNode(nUp);
		}
		return result->next;
	}
	void test() {
		std::vector<int> val0 = { 4, 5, 6 };
		std::vector<int> val1 = { 9, 0, 3 };

		ListNode* node0 = vecToNode(val0);
		ListNode* node1 = vecToNode(val1);
		ListNode* result = addTwoNumbers(node0, node1);
		printListNode(result);
	}
};

File: solutions/python3/problem2.py
# -*- coding: utf-8 -*-
"""
2. Add Two Numbers

You are given two non-empty linked lists representing two non-negative integers.
The digits are stored in reverse order and each of their nodes contain a single digit.
Add the two numbers and return it as a linked list.

You may assume the two numbers do not contain any leading zero, except the number 0 itself.
"""


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(self, l1, l2):
        return self.num_to_list(self.list_to_num(l1) + self.list_to_num(l2))

    def num_to_list(self, num):
        first_node = ListNode(next=ListNode())
        last_node = first_node
        while num > 0:
            digit = num % 10
            num //= 10
            last_node.next = ListNode(digit)
            last_node = last_node.next

        return first_node.next

    def list_to_num(self, l):
        res = l.val
        count = 0

        while l.next is not None:
            count += 1
            l = l.next
            res += l.val * (10 ** count)

        return res